fix(lang/rust): link to old WIT, prose & interface use example (#222)
bytecodealliance/component-docs: component-model/src/language-support/rust.md

## 2. Scaffold a Component with `cargo component`

Create a Rust library that implements the `add` function in the [`adder` world][adder-wit].

First scaffold a project:

```sh

> $ cargo component new adder --lib && cd adder

Reviewer: I believe if you change this you'll need to change lines 96 and 119, as the .wasm file name will change?

Reviewer 2: Done! thanks for the note -- updated the wasm file names

Author: Done! thanks for the note -- updated the wasm file name!

## Exporting an interface with `cargo component`

> While it is possible to export a raw function, prefer exporting an interface, similar to the [sample `adder` world][add-wit]. 

Reviewer 3: ```suggestion
While it is possible to export a raw function, it is a best practice to export an interface, similar to the [sample `adder` world][add-wit]. 
```

> While it is possible to export a raw function, prefer exporting an interface, similar to the [sample `adder` world][add-wit]. 

Reviewer 3: I am not sure I follow why you made these changes to the "Exporting an interface section". Is there a reason you wanted to add in the concept of exporting types too? I may make sense to clarify that to the reader. Maybe: In this section we will export a function and the types it uses to be consumed by another component. 

The way it is written currently, it seems like adding the operand type is related to the general concept of exporting an interface

Author: Ah, so my thinking here was that the above code is *already* exporting the interface, so there was no reason to call that out specifically in a section.

I originally deleted this section, then thought it would be a waste of a good spot to put an example, so I leaned in to the note about "capture a set of functions and types that tend to go together", and wrote an example about that.

> The way it is written currently, it seems like adding the operand type is related to the general concept of exporting an interface

Did you mean "unrelated" here?

Reviewer 2: OK, so this section got removed (see [discussion here](https://github.com/bytecodealliance/component-docs/pull/222#discussion_r2015010725)), should be fine now!

Author: OK, so this section got removed (see https://github.com/bytecodealliance/component-docs/pull/222#discussion_r2015010725), should be fine now!

This often makes it easier to comply with an existing specification or to capture a set of functions and types 
that tend to go together. 

For example, consider the following changes to the `add` interface in the `adder` world:

```wit
package docs:adder@0.1.0;

interface add {

>     variant operand { 

Reviewer: For myself, I'd prefer to see this introduced in a more granular, step-by-step manner.  This section is showing how to move from exporting a loose function to exporting an interface.  Let's try to keep the rest of it the same so the reader can easily see "ah this is what has changed in the WIT and this is how it changes the Rust."  To me, the new stuff about operand conversion gets in the way of that.

I am totally up for showing types in interfaces and how they surface in Rust but for myself I'd be inclined to split that across two stages.  "Now you've seen an interface, let's make the interface FANCY-SCHMANCY."

Reviewer 2: Thanks for the feedback -- I guess for now, I'll remove these changes, it is clearly too much to introduce at once!

Reviewer: @vados-cosmonic From the diff it looks like you've deleted the section rather than just reverting the changes?  Sorry for the miscommunication, that's not what I was trying to suggest.  Or am I misreading?

Author: Ah yes, that's right -- that section actually doesn't make sense anymore, because we *always* export an interface (rather than a function directly)

Reviewer: Oh!  I completely missed that!  Yes, then there is definitely no need for it.  (And if I'd realised the page had already done interfaces, I wouldn't have been bothered by adding the operand stuff... the section title implied this was the first introduction of "how to export an interface" and I didn't double check.  I'm sorry about that!)

        unsigned32(u32)
        signed32(s32)
        zero,
    }

    add: func(x: operand, y: operand) -> result<operand, string>;
}

world adder {
    export add;
}
```

This would be implemented with the following Rust code:

```rust
#[allow(warnings)]
mod bindings;

use bindings::exports::docs::adder::add::{Guest, Operand};

struct Component;

impl Guest for Component {
    fn add(x: Operand, y: Operand) -> Result<Operand, String> {
        let x = convert_operand(x);
        let y = convert_operand(y);
        match x + y {
            v if v == 0 => Ok(Operand::Zero),
            v if v < 0 => i32::try_from(v)
                .map_err(|e| format!("unexpectedly invalid u32: {e}"))
                .map(Operand::Signed32),
            v => u32::try_from(v)
                .map_err(|e| format!("unexpectedly invalid u32: {e}"))
                .map(Operand::Unsigned32),
        }
    }
}

fn convert_operand(operand: Operand) -> i64 {
    match operand {
        Operand::Unsigned32(v) => v as i64,
        Operand::Signed32(v) => v as i64,
        Operand::Zero => 0,
    }
}

bindings::export!(Component with_types_in bindings);
```

> While the code above is unlikely to appear in real interfaces, but it shows the more common case of an interface 

Reviewer 3: ```suggestion
The code above is unlikely to appear in real interfaces; however, it shows the more common case of an interface 
```
